wolfcrypt/src/asn.c

```diff
@@ -8228,52 +8235,73 @@ int ParseCertRelative(DecodedCert* cert, int type, int verify, void* cm)
+             * 1) Is CA boolean set?
+             *      No  - SKIP CHECK
+             *      Yes - Check key usage
+             * 2) Is Key usage extension present?
+             *      No  - goto 3
+             *      Yes - check keyCertSign assertion
+             *     2.a) Is keyCertSign asserted?
+             *          No  - goto 4
+             *          Yes - goto 3
+             * 3) Is pathLen set?
+             *      No  - goto 4
+             *      Yes - check pathLen against maxPathLen.
+             *      3.a) Is pathLen less than maxPathLen?
+             *           No - goto 4
+             *           Yes - set maxPathLen to pathLen and EXIT
+             * 4) Is maxPathLen > 0?
+             *      Yes - Reduce by 1
+             *      No  - ERROR
+             */
 
-            if (cert->ca) {
-                /* Check if cert is CA type and signer has path length set */
-                if (cert->isCA && cert->ca->pathLengthSet) {
-                #if defined(WOLFSSL_WPAS) && !defined(WOLFSSL_NO_ASN_STRICT)
-                    /* WPA Supplicant - has test case that expects self-signed
-                        root CA to have path length == 0 */
-                    if (cert->selfSigned) {
-                        if (cert->ca->pathLength != 0) {
-                            WOLFSSL_MSG("Root CA with path length > 0");
-                            return ASN_PATHLEN_INV_E;
-                        }
-                    }
-                #endif
-                    /* Check if signer is root CA (self-signed) */
-                    if (cert->ca->selfSigned) {
-                        /* Root CA as signer:
-                         * Must have path length > 0 to sign another CA
-                         * If path length == 0 can only sign an end entity
-                         *   certificate, not intermediate CA
-                         */
-                        if (cert->ca->pathLength == 0) {
-                           WOLFSSL_MSG("Root CA with path length == 0");
-                           return ASN_PATHLEN_INV_E;
-                        }
-                    }
-                    else {
-                        /* Intermediate CA signing Intermediate CA */
-                        /* Check path lengths are valid between two CA's */
-                        if (cert->ca->pathLength == 0) {
-                            WOLFSSL_MSG("CA with path length 0 signing a CA");
-                            return ASN_PATHLEN_INV_E;
-                        }
-                        else if (cert->pathLength >= cert->ca->pathLength) {
-                            WOLFSSL_MSG("CA signing CA with longer path length");
-                            return ASN_PATHLEN_INV_E;
-                        }
-                    }
+            if (cert->isCA) {
+                WOLFSSL_MSG("\tCA boolean set, checking pathLen constraint");
+                if (cert->extKeyUsageSet) {
```

Comment:
This could be simplified:
    if (cert->extKeyUsageSet && (cert->extKeyUsage & KEYUSE_KEY_CERT_SIGN) == 0)
        reduceMaxPathByOne = 1;
   else
         checkPathLen = 1;

Agree?

Reply:
I kept it explicit on purpose. Often times we obfuscate things through simplicity and this can lead to the scenario I am addressing where we fall out of compliance. The way I have designed the logic one should be able to read the RFC and see exactly what is being checked and why.